platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java

```diff
@@ -219,4 +219,61 @@ public Set<LocalVariantMetadata> getArtifactVariants() {
             return artifactSets;
         }
     }
+
+    /**
+     * Identifies an artifact set where the owning component identifier has been overridden.
+     * <p>
+     * When overriding the owning component identifier of an artifact, their identity changes. As such,
+     * the identity of the artifact set must also change. This identifier tracks the identity of an
+     * artifact set where the owning component identifier has been overridden, so that caches
+     * properly track the artifacts within this artifact set have an overridden component ID.
+     *
+     * @see org.gradle.internal.resolve.resolver.ResolvedVariantCache
+     */
+    private static class OverrideComponentIdArtifactSetIdentifier implements VariantResolveMetadata.Identifier {
+
+        private final ComponentIdentifier overrideComponentId;
+        private final VariantResolveMetadata.Identifier delegate;
+
+        public OverrideComponentIdArtifactSetIdentifier(
+            ComponentIdentifier overrideComponentId,
+            VariantResolveMetadata.Identifier delegate
+        ) {
+            this.overrideComponentId = overrideComponentId;
+            this.delegate = delegate;
+        }
+
+        @Nullable
+        public static OverrideComponentIdArtifactSetIdentifier of(
+            ComponentIdentifier overrideComponentId,
+            @Nullable VariantResolveMetadata.Identifier delegate
+        ) {
+            if (delegate == null) {
+                return null;
+            }
```

Comment:
🤔 Instead of having null input return null results here, it might be better to just null check prior to calling `of`.

Reply:
The whole point of `of` is to deal with the nullable ID

Reply:
I agree, but see that as a good thing?

Reply:
Why have a method that just calls new and passes its two parameters verbatim to a constructor? Seems a bit worthless to me

Reply:
Yeah this could probably use a better name.

It's just more concise in the calling method where you either need a ternary, which I didn't like initially, or a to assign the new id to a variable in an if, which was more wordy than I wanted

In any case this is just a style thing and is fine the way it is imo